sklearn/ensemble/_forest.py

```diff
@@ -2633,7 +2634,12 @@ def fit_transform(self, X, y=None, sample_weight=None):
         super().fit(X, y, sample_weight=sample_weight)
 
         self.one_hot_encoder_ = OneHotEncoder(sparse=self.sparse_output)
-        return self.one_hot_encoder_.fit_transform(self.apply(X))
+        X_leaf = self.apply(X)
+        self.one_hot_encoder_.fit(X_leaf)
+        self._n_features_out = np.product(
+            np.asarray(self.one_hot_encoder_.categories_).shape
+        )
+        return self.one_hot_encoder_.transform(X_leaf)
```

Comment:
I think we can keep the original `fit_transform` and do a little bit of logic after:

```python
output = self.one_hot_encoder_.fit_transform(self.apply(X))
self._n_features_out = output.shape[1]
return output
```

(Also there is a `numpy.VisibleDeprecationWarning` when using `asarray` on a ragged nested list)